examples/gcoap/gcoap_cli.c

```diff
@@ -93,11 +96,39 @@ static void _resp_handler(unsigned req_state, coap_pkt_t* pdu,
  */
 static ssize_t _stats_handler(coap_pkt_t* pdu, uint8_t *buf, size_t len)
 {
-    gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+    /* read coap method type in packet */
+    unsigned method_flag = coap_method2flag(coap_get_code_detail(pdu));
+
+    switch(method_flag) {
+        case COAP_GET:
+            gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+
+            /* write the response buffer with the request count value */
+            size_t payload_len = fmt_u16_dec((char *)pdu->payload, req_count);
+
+            return gcoap_finish(pdu, payload_len, COAP_FORMAT_TEXT);
+        case COAP_PUT:
+        {
+            /* Read and convert the content of the payload */
+            req_count = (uint16_t)strtoul((char*)pdu->payload, NULL, 10);
```

Comment:
strtoul() expects the first argument to be a null-terminated string. pdu->payload is not guaranteed to be null-terminated. This is why the PUT/POST handler from nanocoap example created the temporary buffer.

Below is my solution, which worked for me on native. I added the bad request response because a payload that is too long is nonsense, and it also illustrates how easy it is to send an error response. You could also say that any non-zero value is nonsense, but I didn't go that far. :-)

Also, you don't need to send a payload with the success response. If the response is a 2.04, the sender knows they succeeded.

```
        case COAP_PUT:
            /* convert the payload to an integer and update the internal value */
            if (pdu->payload_len <= 5) {
                char payload[6] = { 0 };
                memcpy(payload, (char*)pdu->payload, pdu->payload_len);

                req_count = (uint16_t)strtoul(payload, NULL, 10);

                return gcoap_response(pdu, buf, len, COAP_CODE_CHANGED);
            }
            else {
                return gcoap_response(pdu, buf, len, COAP_CODE_BAD_REQUEST);
            }

```

I would also update the comment for this handler function to make the handling clear.

```
/*
 * Server callback for /cli/stats. Accepts either a GET or a PUT.
 *
 * GET: Returns the count of packets sent by the CLI.
 * PUT: Updates the count of packets. Rejects an obviously bad request, but
 *      allows any two byte value for example purposes. Semantically, the only
 *      valid action is to set the value to 0.
 */

```